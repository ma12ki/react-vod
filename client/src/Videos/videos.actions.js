const prefix = '[VIDEOS]';

export const LIST_LOAD_START = `${prefix}LIST_LOAD_START`;
export const LIST_LOAD_SUCCESS = `${prefix}LIST_LOAD_SUCCESS`;
export const LIST_LOAD_ERROR = `${prefix}LIST_LOAD_ERROR`;

export const listLoadStart = () => ({ type: LIST_LOAD_START });
export const listLoadSuccess = (payload) => ({ type: LIST_LOAD_SUCCESS, payload });
export const listLoadError = (payload) => ({ type: LIST_LOAD_ERROR, payload });
