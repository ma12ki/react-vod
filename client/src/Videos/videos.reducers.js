import {
    LIST_LOAD_START,
    LIST_LOAD_SUCCESS,
    LIST_LOAD_ERROR,
} from './videos.actions';

const defaultState = {
    loading: false,
    items: [],
    error: null,
};

export const videos = (state = defaultState, action = {}) => {
    switch (action.type) {
        case LIST_LOAD_START: {
            return {
                ...state,
                loading: true,
                error: null,
            };
        }
        case LIST_LOAD_SUCCESS: {
            return {
                ...state,
                loading: false,
                items: action.payload,
                error: null,
            };
        }
        case LIST_LOAD_ERROR: {
            return {
                ...state,
                loading: false,
                error: action.payload,
            };
        }
        default:
            return state;
    }
}