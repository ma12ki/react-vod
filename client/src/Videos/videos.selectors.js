const videos = (state) => state.videos;
const loading = (state) => videos(state).loading;
const error = (state) => videos(state).error;
const items = (state) => videos(state).items || {};
const videosResult = (state) => items(state).result || [];
const entities = (state) => items(state).entities || {};
const videosEntities = (state) => entities(state).videos || {};

const videosItems = (state) => {
    const entities = videosEntities(state);
    const result = videosResult(state);

    return result.map((id) => entities[id]);
};

export {
    loading,
    error,
    videosEntities,
    videosResult,
    videosItems,
};
