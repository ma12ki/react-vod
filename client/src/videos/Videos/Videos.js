import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { getInitialized } from '../selectors';
import { loadVideosStart } from '../ducks';
import { VideoList } from '../VideoList';
import { VideoSearch } from '../VideoSearch';
import { VideoRefresh } from '../VideoRefresh';
import styles from './videos.module.css';

export class Videos extends React.PureComponent {
    static propTypes = {
        initialized: PropTypes.bool.isRequired,
        loadVideos: PropTypes.func.isRequired,
    }

    componentDidMount() {
        if (!this.props.initialized) {
            this.props.loadVideos();
        }
    }

    render() {
        return (
            <div>
                <div className={styles.flex}>
                    <VideoSearch />
                    <VideoRefresh />
                </div>
                <VideoList />
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    initialized: getInitialized(state),
});

const mapDispatchToProps = (dispatch) => ({
    loadVideos: () => dispatch(loadVideosStart())
});

export default connect(mapStateToProps, mapDispatchToProps)(Videos);
