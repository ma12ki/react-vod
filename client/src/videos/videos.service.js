import { schema, normalize } from 'normalizr';

import { http } from '../utils';

const videosSchema = new schema.Entity('videos');

const mapVideoDtoToModel = (videoDto) => {
    return {
        ...videoDto,
        dateCreated: new Date(videoDto.dateCreated),
        dateModified: new Date(videoDto.dateModified),
    };
};

const mapVideoDtosToModel = (videoDtos) => {
    return videoDtos.map((videoDto) => mapVideoDtoToModel(videoDto));
};

const getVideoList$ = () => {
    return http.get$('videos')
        .delay(1000)
        .map((videoDtos) => {
            const videos = mapVideoDtosToModel(videoDtos);
            return normalize(videos, [videosSchema]);
        });
};

const refreshVideoList$ = () => {
    return http.get$('videos/refresh')
        .delay(1000)
        .map((videoDtos) => {
            const videos = mapVideoDtosToModel(videoDtos);
            return normalize(videos, [videosSchema]);
        });
};

const getOneVideo$ = (id) => {
    return http.get$(`videos/${id}`)
        .delay(1000)
        .map((videoDto) => {
            const video = mapVideoDtoToModel(videoDto);
            return normalize(video, videosSchema);
        });
};

export {
    getVideoList$,
    refreshVideoList$,
    getOneVideo$,
};
