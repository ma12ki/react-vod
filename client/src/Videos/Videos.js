import React from 'react';
import { connect } from 'react-redux';

import { listLoadStart } from './videos.actions';

import VideoList from './VideoList/VideoList';

const Videos = (props) => {
    return (
        <div>
            <button onClick={props.onLoad} type='button'>load</button>
            <VideoList />
        </div>
    );
};

const mapStateToProps = (state) => ({
    items: [],
});

const mapDispatchToProps = (dispatch) => ({
    onLoad: () => dispatch(listLoadStart()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Videos);
