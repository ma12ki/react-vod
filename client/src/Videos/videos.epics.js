import { Observable } from 'rxjs';

import {
    LIST_LOAD_START,
    listLoadSuccess,
    listLoadError,
} from './videos.actions';
import { getVideoList$ } from './videos.service';

export const loadVideosList$ = action$ =>
    action$.ofType(LIST_LOAD_START)
        .switchMap(() => getVideoList$())
            .map((res) => listLoadSuccess(res))
            .catch((err) => Observable.of(listLoadError(err)));
