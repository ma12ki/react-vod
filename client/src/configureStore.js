import { connectRoutes } from 'redux-first-router';
import { combineReducers, createStore, applyMiddleware, compose } from 'redux';
import { createEpicMiddleware } from 'redux-observable';
import createHistory from 'history/createBrowserHistory';

import { routesMap } from './routes';

import { videos } from './Videos/videos.reducers';
import { loadVideosList$ } from './Videos/videos.epics';

const history = createHistory();
const { reducer, middleware, enhancer } = connectRoutes(history, routesMap); 

const epicMiddleware = createEpicMiddleware(loadVideosList$);

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const rootReducer = combineReducers({ location: reducer, videos });
const middlewares = applyMiddleware(middleware, epicMiddleware);
const store = createStore(rootReducer, composeEnhancers(enhancer, middlewares));

export {
    store,
};
