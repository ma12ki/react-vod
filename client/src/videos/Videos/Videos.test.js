import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { shallow, mount } from 'enzyme';

import { testPropTypes } from '../../testHelpers';
import { Videos } from './Videos';

const mockState = {
    location: {
        routesMap: {}
    },
    videos: {}
};
const makeWrapper = ({ initialized = false, loadVideos = () => null } = {}) => shallow(<Videos initialized={initialized} loadVideos={loadVideos} />);
const mountWrapper = ({ initialized = false, loadVideos = () => null, mockStore = configureStore()(() => mockState) } = {}) => {
    return mount(
        <Provider store={mockStore}>
            <Videos initialized={initialized} loadVideos={loadVideos} />
        </Provider>
    );
};

describe('Video', () => {

    it('matches the snapshot', () => {
        const wrapper = makeWrapper();

        expect(wrapper).toMatchSnapshot();
    });

    it('requires "initialized" bool', () => {
        testPropTypes(
            Videos,
            'initialized',
            [true, false],
            [123, 'abc', null, undefined],
            { loadVideos: () => null }
        );
    });

    it('requires "loadVideos" function', () => {
        testPropTypes(
            Videos,
            'loadVideos',
            [() => null],
            [123, 'abc', null, undefined],
            { initialized: true }
        );
    });

    it('calls loadVideos if "initialized" prop is FALSE', () => {
        const loadVideos = jest.fn();
        mountWrapper({ initialized: false, loadVideos });

        expect(loadVideos).toHaveBeenCalled();
    });

    it('does NOT call loadVideos if "initialized" prop is TRUE', () => {
        const loadVideos = jest.fn();
        mountWrapper({ initialized: true, loadVideos });

        expect(loadVideos).not.toHaveBeenCalled();
    });

});