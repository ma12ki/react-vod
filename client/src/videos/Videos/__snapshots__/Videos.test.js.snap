// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video matches the snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <div
        className="flex"
    >
        <Connect(VideoSearch) />
        <Connect(Component) />
    </div>
    <Connect(Component) />
</div>,
  "nodes": Array [
    <div>
      <div
            className="flex"
      >
            <Connect(VideoSearch) />
            <Connect(Component) />
      </div>
      <Connect(Component) />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 1,
      "_context": Object {},
      "_currentElement": <Videos
        initialized={false}
        loadVideos={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Videos {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "initialized": false,
          "loadVideos": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <div
                    className="flex"
          >
                    <Connect(VideoSearch) />
                    <Connect(Component) />
          </div>
          <Connect(Component) />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <div
                    className="flex"
          >
                    <Connect(VideoSearch) />
                    <Connect(Component) />
          </div>
          <Connect(Component) />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Videos
    initialized={false}
    loadVideos={[Function]}
/>,
}
`;
